### ISSUE TITLE
Calling window.inner_size inside event callback on Wayland causes deadlock

### ISSUE THREAD

**Issue by alec-deason** [rust-windowing/winit#1417]
`platform_impl::linux::wayland::window::WindowStore::for_each` holds open a lock on the window's size which causes any call to window.inner_size or other functions which access that data to deadlock when called from inside the event loop callback. This is a regression between 0.20.0-alpha4 and 0.20.0 but I'm not sure exactly which change did it.

The problematic line is here: https://github.com/rust-windowing/winit/blob/master/src/platform_impl/linux/wayland/window.rs#L486

Here is a minimal example that reproduces the issue for me. The dependency on https://github.com/parasyte/pixels is there just to draw something and force Wayland to actually create a window.

```
use pixels::{wgpu::Surface, Error, Pixels, SurfaceTexture};
use winit::{
    event::{Event, WindowEvent},
    event_loop::{ControlFlow, EventLoop},
    window::WindowBuilder,
};

fn main() {
    let event_loop = EventLoop::new();

    let window = WindowBuilder::new()
        .with_title("A fantastic window!")
        .with_inner_size(winit::dpi::LogicalSize::new(128.0, 128.0))
        .build(&event_loop)
        .unwrap();

    let size = window.inner_size();
    let mut pixels = {
        let surface = Surface::create(&window);
        let surface_texture = SurfaceTexture::new(size.width, size.height, surface);
        Pixels::new(size.width, size.height, surface_texture).unwrap()
    };

    window.request_redraw();

    event_loop.run(move |event, _, control_flow| {
        *control_flow = ControlFlow::Wait;
        println!("{:?}", event);


        match event {
            Event::RedrawRequested { .. } => {
                pixels.render();
            }
            Event::WindowEvent {
                event: WindowEvent::Resized(new_size),
                ..
            } => {
                eprintln!("Starting resize event");
                eprintln!("new size: {:?}", window.inner_size());
                eprintln!("Done with resize event");
            }
            Event::WindowEvent {
                event: WindowEvent::CloseRequested,
                window_id,
            } if window_id == window.id() => *control_flow = ControlFlow::Exit,
            Event::MainEventsCleared => {
                window.request_redraw();
            }
            _ => (),
        }
    });
}
```

And the backtrace from running that:

```
0x00007ffff7f7041f in __lll_lock_wait () from /usr/lib/libpthread.so.0
(gdb) bt
#0  0x00007ffff7f7041f in __lll_lock_wait () from /usr/lib/libpthread.so.0
#1  0x00007ffff7f68ea3 in pthread_mutex_lock () from /usr/lib/libpthread.so.0
#2  0x0000555555842775 in std::sys::unix::mutex::Mutex::lock (self=0x555556069500)
    at /rustc/b5a3341f1b8b475990e9d1b071b88d3c280936b4/src/libstd/sys/unix/mutex.rs:57
#3  0x0000555555864931 in std::sys_common::mutex::Mutex::raw_lock (self=0x555556069500)
    at /rustc/b5a3341f1b8b475990e9d1b071b88d3c280936b4/src/libstd/sys_common/mutex.rs:41
#4  0x0000555555865ef5 in std::sync::mutex::Mutex<T>::lock (self=0x555556090730)
    at /rustc/b5a3341f1b8b475990e9d1b071b88d3c280936b4/src/libstd/sync/mutex.rs:218
#5  0x00005555556e91d0 in winit::platform_impl::platform::wayland::window::Window::inner_size (
    self=0x7fffffffc8a8)
    at /home/alec/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.20.0/src/platform_impl/linux/wayland/window.rs:225
#6  0x000055555564d529 in winit::platform_impl::platform::Window::inner_size (self=0x7fffffffc8a0)
    at /home/alec/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.20.0/src/platform_impl/linux/mod.rs:276
#7  0x0000555555616e66 in winit::window::Window::inner_size (self=0x7fffffffc8a0)
    at /home/alec/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.20.0/src/window.rs:479
#8  0x0000555555616892 in winit_example::main::{{closure}} (event=..., control_flow=0x7fffffffbab0)
    at src/main.rs:40
#9  0x0000555555616653 in core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut (self=0x7fffffffb690, args=...)
    at /rustc/b5a3341f1b8b475990e9d1b071b88d3c280936b4/src/libcore/ops/function.rs:274
#10 0x000055555564038d in winit::platform_impl::platform::sticky_exit_callback (evt=...,
    target=0x7fffffffcc58, control_flow=0x7fffffffbab0, callback=0x7fffffffb690)
    at /home/alec/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.20.0/src/platform_impl/linux/mod.rs:697
#11 0x00005555556228d9 in winit::platform_impl::platform::wayland::event_loop::EventLoop<T>::post_dispatch_triggers::{{closure}} (event=...)
    at /home/alec/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.20.0/src/platform_impl/linux/wayland/event_loop.rs:684
#12 0x0000555555622b46 in winit::platform_impl::platform::wayland::event_loop::EventLoop<T>::post_dispatch_triggers::{{closure}} (window=...)
    at /home/alec/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.20.0/src/platform_impl/linux/wayland/event_loop.rs:722
#13 0x000055555565d904 in winit::platform_impl::platform::wayland::window::WindowStore::for_each (
    self=0x55555605f970, f=...)
    at /home/alec/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.20.0/src/platform_impl/linux/wayland/window.rs:484
#14 0x00005555556227a4 in winit::platform_impl::platform::wayland::event_loop::EventLoop<T>::post_dispatch_triggers (self=0x7fffffffcb98, callback=0x7fffffffc828, control_flow=0x7fffffffbab0)
    at /home/alec/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.20.0/src/platform_impl/linux/wayland/event_loop.rs:704
#15 0x0000555555621da3 in winit::platform_impl::platform::wayland::event_loop::EventLoop<T>::run_return (
    self=0x7fffffffcb98, callback=...)
    at /home/alec/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.20.0/src/platform_impl/linux/wayland/event_loop.rs:485
#16 0x0000555555622f5a in winit::platform_impl::platform::wayland::event_loop::EventLoop<T>::run (
    self=..., callback=...)
    at /home/alec/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.20.0/src/platform_impl/linux/wayland/event_loop.rs:445
#17 0x000055555563ff11 in winit::platform_impl::platform::EventLoop<T>::run (self=..., callback=...)
    at /home/alec/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.20.0/src/platform_impl/linux/mod.rs:643
#18 0x000055555564a643 in winit::event_loop::EventLoop<T>::run (self=..., event_handler=...)
    at /home/alec/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.20.0/src/event_loop.rs:148
```

I'm on Arch Linux and using Sway 1.2-d19f4f7b

**Comment by alec-deason** [comment]
I'm new to this codebase so I'm a bit hazy on how exactly to approach a fix for this. Possibly something like what was done in [this pr](https://github.com/rust-windowing/winit/pull/1369). Or maybe `EventLoop::post_dispatch_triggers` could acquire and release the lock each time it actually made a change.

**Comment by goddessfreya** [comment]
#1369 should have no effect on the Wayland backend.

**Comment by alec-deason** [comment]
Oh, I wasn't suggesting that it was related just that the way that problem was fixed might also work for this problem.

**Comment by alec-deason** [comment]
I have some un-clarity about the semantics of set_inner_size. It does two things: 1) it calls `self.frame.lock().unwrap().resize(w, h)` which requests that the Wayland server resize the window. 2) it sets the value of `self.size` which acts as a local cache of the current size of the window and is the thing protected by the mutex that is deadlocking.

That's incorrect (unless I'm misunderstanding). The Wayland server is not obligated to respect the `resize` request and in the case of tiling compositors like Sway it usually won't. So `set_inner_size()` doesn't always change the size of the window but it does change the value reported by `inner_size()`. That means that in cases where Wayland ignores the resize request `inner_size()` will return an incorrect result until it's cache is un-poisoned by the arrival of a new `Resized` event.

Based on that, I propose this fix: `set_inner_size()` no longer updates the cached size which now can only be updated by the arrival of a `Resize` event and thus no longer needs it's own mutex. With the mutex gone it can also be safely be read from inside the callback. Also `inner_size()` will return the correct value because it will always contain the size most recently reported by the Wayland server via `Resize` events.

This isn't a perfect fix because a call to `set_inner_size()` or anything else that needs the mutex on `frame` will still deadlock from inside the callback.

I need to go read up on smithay_client_toolkit and Wayland to make sure I'm clear on what's actually happening then I'll write this up.

Also there seems to be an unrelated bug in that the call to `resize()` inside `set_inner_size()` is not followed by a call to `refresh()`. I think there's an easy fix for that by just setting `needs_refresh`. I'll write a bug/PR for that as well once I've gone over the docs and made sure I'm not just confused.

**Comment by alec-deason** [comment]
Ugh, so I was super wrong about a bunch of stuff both in `winit` and in Wayland. The caching behavior of `set_inner_size()` still seems suspicious to me but I no longer think I understand what's happening well enough to suggest a change.

**Comment by farnoy** [comment]
I have a related problem, should I create a new issue? For me, it hangs under X11, in `Window::scale_factor()`:

My code is (simplified):

```rust
self.events_loop
    .run_return(|event, window_target, control_flow| {
        dbg!(&event);
        match event {
            Event::WindowEvent {
                event: WindowEvent::Resized(PhysicalSize { width, height }),
                ..
            } => {
                println!("The window was resized to {}x{}", width, height);
                let logical_size = PhysicalSize { width, height }.to_logical::<f32>(window.scale_factor());
                println!("logical {:?}", logical_size);
                resized = true;
            }
            Event::WindowEvent {
                event: WindowEvent::ScaleFactorChanged { scale_factor, .. },
                ..
            } => {
                println!("Scale factor changed {}", scale_factor);
            }
            _ => (),
        };
        *control_flow =winit::event_loop::ControlFlow::Exit;
    });

```


```
[src/ecs/systems.rs:104] &event = NewEvents(
    Init,
)
[src/ecs/systems.rs:104] &event = WindowEvent {
    window_id: WindowId(
        X(
            WindowId(
                144703489,
            ),
        ),
    ),
    event: Resized(
        PhysicalSize {
            width: 958,
            height: 1055,
        },
    ),
}
The window was resized to 958x1055
Process 88288 stopped
* thread #1, name = 'renderer', stop reason = signal SIGSTOP
    frame #0: 0x00007ffff7c74e9d libc.so.6`syscall + 29
libc.so.6`syscall:
->  0x7ffff7c74e9d <+29>: cmpq   $-0xfff, %rax             ; imm = 0xF001
    0x7ffff7c74ea3 <+35>: jae    0x7ffff7c74ea6            ; <+38>
    0x7ffff7c74ea5 <+37>: retq
    0x7ffff7c74ea6 <+38>: movq   0xc5fb3(%rip), %rcx       ;  + 688
(lldb) bt
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
* thread #1, name = 'renderer', stop reason = signal SIGSTOP
  * frame #0: 0x00007ffff7c74e9d libc.so.6`syscall + 29
    frame #1: 0x00005555561a1c40 renderer`parking_lot_core::thread_parker::imp::ThreadParker::futex_wait::hcd44f979e0b9648b(self=0x00007ffff7a30768, ts=Option<libc::unix::timespec> @ 0x00007fffffff3390) at linux.rs:111:12
    frame #2: 0x00005555561a195a renderer`_$LT$parking_lot_core..thread_parker..imp..ThreadParker$u20$as$u20$parking_lot_core..thread_parker..ThreadParkerT$GT$::park::h8d19677504047567(self=0x00007ffff7a30768) at linux.rs:65:12
    frame #3: 0x00005555561a1169 renderer`parking_lot_core::parking_lot::park::_$u7b$$u7b$closure$u7d$$u7d$::h1bdac6f6598c56bd(thread_data=0x00007ffff7a30748) at parking_lot.rs:611:16
    frame #4: 0x00005555561a0de6 renderer`parking_lot_core::parking_lot::park::h59fb06cec64a9e9c at parking_lot.rs:183:4
    frame #5: 0x00005555561a0d08 renderer`parking_lot_core::parking_lot::park::h59fb06cec64a9e9c(key=93825010196944, validate=closure-0 @ 0x00007fffffff3568, before_sleep=closure-1 @ 0x00007fffffff3570, timed_out=closure-2 @ 0x00007fffffff3578, park_token=(__0 = 0), timeout=<unavailable>) at parking_lot.rs:576
    frame #6: 0x00005555561a00ac renderer`parking_lot::raw_mutex::RawMutex::lock_slow::h4c5b6fd13372cc50(self=0x00005555566761d0, timeout=Option<std::time::Instant> @ 0x00007fffffff3720) at raw_mutex.rs:256:16
    frame #7: 0x0000555555f34369 renderer`_$LT$parking_lot..raw_mutex..RawMutex$u20$as$u20$lock_api..mutex..RawMutex$GT$::lock::h7ee3081a99910748(self=0x00005555566761d0) at raw_mutex.rs:72:12
    frame #8: 0x0000555555ea0fa6 renderer`lock_api::mutex::Mutex$LT$R$C$T$GT$::lock::h8edf7bfe66e72c0c(self=0x00005555566761d0) at mutex.rs:156:8
    frame #9: 0x000055555568e80b renderer`winit::platform_impl::platform::x11::window::UnownedWindow::current_monitor::h1bc228d5a1846ed2(self=0x00005555566761a0) at window.rs:746:8
    frame #10: 0x000055555568e6db renderer`winit::platform_impl::platform::x11::window::UnownedWindow::scale_factor::hc870ddc7a5871487(self=0x00005555566761a0) at window.rs:1276:8
    frame #11: 0x00005555557b4965 renderer`winit::platform_impl::platform::Window::scale_factor::hb2a5e799c682c28f(self=0x0000555556ac90e0) at mod.rs:347:33
    frame #12: 0x00005555556f79b6 renderer`winit::window::Window::scale_factor::h5eadc5ee2c024ae9(self=0x0000555556ac90e0) at window.rs:387:8
    frame #13: 0x00005555557f31c4 renderer`renderer::ecs::systems::InputHandler::exec::_$u7b$$u7b$closure$u7d$$u7d$::h495f5e38a5cd9028(event=Event<()> @ 0x00007fffffff3d80, window_target=0x0000555556672c80, control_flow=0x00007fffffff3ec0) at systems.rs:113:92
```

**Comment by goddessfreya** [comment]
That would be a different bug, please file separately, @farnoy.

**Comment by dvc94ch** [comment]
This is the simplest fix, but we might prefer to use the `crossbeam` `AtomicCell`, since all these locks look like they're bound to give more trouble.

```rust
diff --git a/src/platform_impl/linux/wayland/window.rs b/src/platform_impl/linux/wayland/window.rs
index 542fb068..6976093d 100644
--- a/src/platform_impl/linux/wayland/window.rs
+++ b/src/platform_impl/linux/wayland/window.rs
@@ -492,9 +492,10 @@ impl WindowStore {
         for window in &mut self.windows {
             let opt_arc = window.frame.upgrade();
             let mut opt_mutex_lock = opt_arc.as_ref().map(|m| m.lock().unwrap());
+            let mut size = { *window.size.lock().unwrap() };
             f(WindowStoreForEach {
                 newsize: window.newsize.take(),
-                size: &mut *(window.size.lock().unwrap()),
+                size: &mut size,
                 prev_dpi: window.current_dpi,
                 new_dpi: window.new_dpi,
                 closed: window.closed,
@@ -503,6 +504,7 @@ impl WindowStore {
                 wid: make_wid(&window.surface),
                 frame: opt_mutex_lock.as_mut().map(|m| &mut **m),
             });
+            *window.size.lock().unwrap() = size;
             if let Some(dpi) = window.new_dpi.take() {
                 window.current_dpi = dpi;
             }
```

**Comment by vberger** [comment]
@alec-deason 

> 1) it calls self.frame.lock().unwrap().resize(w, h) which requests that the Wayland server resize the window.

No, that is not what this method does. The Wayland server does not resize the window, the size of the window is defined by the size of the contents that are drawn on it. This method call resizes the decorations drawn by SCTK when server-side decorations are not available. This is the main reason winit keeps a cached value of the size of its window: nothing else will keep track of it.

Now, for this deadlock, this seems to be the result of an unfortunate evolution of the codebase. the `for_each` method was not initially design for user callbacks to be invoked from within it. And the new API was fitted into it a posteriori without getting this correctly, I guess.

@dvc94ch You patch (and PR #1438 ) would indeed solve the deadlock, and I'm fine with merging it. The real long term solution would of course be a more in-depth refactor of the backend, but I think this would be better done once I release with the next version of SCTK, when it is ready.

**Comment by alec-deason** [comment]
@vberger Yeah, sorry about that. I was very confused about how several things worked.
